lib/spack/spack/solver/asp.py

```diff
@@ -2405,6 +2405,13 @@ def _reusable_specs(self):
                 # TODO: update mirror configuration so it can indicate that the
                 # TODO: source cache (or any mirror really) doesn't have binaries.
                 pass
+
+        # If we only want to reuse dependencies, remove the root specs
+        if self.reuse == "dependencies":
+            reusable_specs = [
+                spec for spec in reusable_specs if not any(spec.satisfies(root) for root in specs)
+            ]
```

Comment:
I think this is redundant, since it includes specs that have a root as a dependency. The solve will rule that out, but potentially we have more hashes than we want as facts. Did you try:
```suggestion
        # If we only want to reuse dependencies, remove the root specs
        if self.reuse == "dependencies":
            reusable_specs = [
                spec for spec in reusable_specs if not any(root in spec for root in specs)
            ]
```
?

Reply:
I might be reading the code wrong but aren't we hitting the anonymous case in `__contains__` anyway, which simply calls `satisfies`: https://github.com/spack/spack/blob/develop/lib/spack/spack/spec.py#L4001-L4002
I would therefore assume that this behaves identically. I also did some tests and couldn't provoke any differences.

Reply:
Reusable specs are all concrete specs collected either from the DB or from buildcaches, root specs can't be anonymous. I checked the ASP before posting. Result is the same, we just have a lot of input in the ASP which is useless. 

Try for instance:
1. Install `hdf5+mpi` (fake installs are good for this test)
2. Create an environment with `hdf5+mpi` and zlib as root specs + `reuse:dependencies`
3. `spack solve --show=asp` for the environment

I think you'll see all the specs in there, but `hdf5` and `zlib` There are specs though that reference the non-existing zlib.

I think there might be other potential pitfalls we might need to check, e.g. if I can reuse the same package as the root but with different variant in an environment. Going back to the example above do we want to reuse `hdf5~mpi` if it is installed, since the the root says `hdf5+mpi`? That matters if we have a root spec that depend on `hdf5`. I didn't test-drive that case yet.

Reply:
I got what you meant now, thanks! I confused myself with the "root specs" in the snippet above and actually tested it outside of an environment (which hits the anonymous spec path).

I now did some tests inside an environment and could reproduce your results. I guess the interaction can be a bit surprising if a common dependency is a root spec: I added a new `cmake` version but since `zlib` is a root spec and thus filtered out, `cmake` gets rebuilt even though it could theoretically be reused. I guess there is no way around that, though.

If I add `enzo` (which requires `hdf5~mpi`) in your example above, everything seems to continue working as expected:
```
==> Error: 'hdf5' required multiple values for single-valued variant 'mpi'
    Requested '~mpi' and '+mpi'
```

I also made the change you suggested and pushed a new version.